### PR TITLE
Completed Apache config template with validation

### PR DESCRIPTION
**Pull request by akmiller01**
The validator was a pain, requiring some additional pieces beyond the original configs.

### CODE REVIEW

File: roles/common/apache2/apache2_install/tasks/main.yml [hunk 1]
```diff
@@ -8,12 +8,40 @@
...
+- name: Apache | Create log folder
+  file:
+    path: /etc/apache2/logs
+    state: directory
     owner: root
     group: root
-    mode: 0644
+    mode: 0775
+    
+# create a test configuration file that links in all of the required config
```
Comment (allthatilk) [inline roles/common/apache2/apache2_install/tasks/main.yml]: This is not as clear as it needs to be, long comments are fine if you need them. This one feels a bit cut off.

File: roles/common/apache2/apache2_install/tasks/main.yml [hunk 2]
```diff
@@ -8,12 +8,40 @@
...
     - "{{ apache2_modules }}"
-
-- name: Copy apache2 config file to sites-available
-  copy:
-    remote_src: no
-    src: "{{ apache2_conf_template }}"
-    dest: "{{ apache2_conf_destination }}"
+    
+# Create log directory
+- name: Apache | Create log folder
```
Comment (allthatilk) [inline roles/common/apache2/apache2_install/tasks/main.yml]: The pipe in this name is a little confusing with the pipes in statements later on

File: roles/common/apache2/apache2_install/tasks/main.yml [hunk 3]
```diff
@@ -8,12 +8,40 @@
...
+     src: apache2.conf.j2 
+     dest: "/tmp/apache2-test.conf"
+     mode: 0644
+     owner: root
+     group: root
+     validate: 'apachectl -t -f %s'
+  register: apache_result
+  ignore_errors: yes
+
+# if the config test succeeds we can put the new vhost live
```
Comment (allthatilk) [inline roles/common/apache2/apache2_install/tasks/main.yml]: can we expand vhost to virtual host?

File: roles/common/apache2/apache2_install/templates/apache2.conf.j2 [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# Include module configuration:
```
Comment (allthatilk) [inline roles/common/apache2/apache2_install/templates/apache2.conf.j2]: Are these first 10 lines necessary? They don't appear to be in the original apache2.conf files.

File: roles/common/apache2/apache2_install/templates/apache2.conf.j2 [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+Include /etc/apache2/mods-enabled/*.load
+Include /etc/apache2/mods-enabled/*.conf
+
+# Include ports listing
+Include /etc/apache2/ports.conf
+
+{% if apache2_servername is defined %}
+ServerName {{apache2_servername}}
```
Comment (allthatilk) [inline roles/common/apache2/apache2_install/templates/apache2.conf.j2]: For the sake of formatting consistency, there needs to be spaces around the variable inside the brackets

File: roles/common/apache2/apache2_install/templates/apache2.conf.j2 [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  ServerAdmin webmaster@localhost
+  
+  {% if apache2_documentroot is defined %}
+  DocumentRoot {{apache2_documentroot}}
+  {% endif %}
+
+  ErrorLog ${APACHE_LOG_DIR}/error.log
+  CustomLog ${APACHE_LOG_DIR}/access.log combined
+
+  {% for directory in apache2_directories %}
```
Comment (allthatilk) [inline roles/common/apache2/apache2_install/templates/apache2.conf.j2:11]: Can you add a comment here to explain what this block is doing. It's pretty straightforward but the nested if statements mean this block needs an overview if refactoring isn't possible.

File: roles/common/apache2/apache2_install/templates/apache2.conf.j2 [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      {% for option in directory.options %}
+    {{option.line}}
+      {% endfor %}
+    {% endif %}
+    Require all granted
+    {% endif %}
+  </Directory>
+  {% endfor %}
+
+  {% for apache2_extra in apache2_extras %}
```
Comment (allthatilk) [inline roles/common/apache2/apache2_install/templates/apache2.conf.j2:31]: what does apache2 extras refer to?

File: roles/datastore/datastore_setup/vars/main.yml [hunk 8]
```diff
@@ -15,6 +15,22 @@ repo_dir: IATI-Datastore/
...
+apache2_port: 80
+
+apache2_servername: localhost
+
+apache2_directories:
+  - root: /
+  - options:
+    - line: Options FollowSymLinks
+    
+apache2_extras:
```
Comment (allthatilk) [inline roles/datastore/datastore_setup/vars/main.yml:29]: The name of this variable needs to be clearer

File: roles/wagtail/wagtail_setup/vars/main.yml [hunk 9]
```diff
@@ -20,3 +18,23 @@ db_password: !vault |
...
+apache2_documentroot: /var/www/html
+
+apache2_directories:
+  - alias_source: /static
+    alias_destination: /home/iati-website/preview-site/iati/staticfiles
+    root: /home/iati-website/preview-site/iati/staticfiles
+  - root: /home/iati-website/preview-site/iati/iati
+    file_location: wsgi.py
+    
+apache2_extras:
```
Comment (allthatilk) [inline roles/wagtail/wagtail_setup/vars/main.yml:35]: same as the comment in the datastore vars file

**Review by allthatilk — CHANGES_REQUESTED** [review]
Nice work! Most things are just around consistent formatting and comment/variable name clarity. I also think we need to have a look at whether this templating functionality needs to be extracted to its own role or not.

**Comment by akmiller01** [comment]
Changes placed into apache-template-feature branch.

**Pull request closed without merging**